pkg/osvscanner/customgitignore/walk_up_to_root.go

```diff
@@ -0,0 +1,234 @@
+// behaviours:
+//
+// - parsing file:
+// 	- find the file's dir and then run the following rules...
+//
+// - parsing dir at the root of a git-repo
+// 	- with recursive flag:
+// 		- read all .gitignore files in repo
+// 		- read .git/info/exclude for repo
+// 	- without recursive flag:
+// 		- read .gitignore files in repo-root
+// 		- read .git/info/exclude for repo
+//
+// - parsing dir within, a git-repo
+// 	- with recursive flag:
+// 		- read .gitignore in start dir
+// 		- read all .gitignore files in child-dirs
+// 		- read all .gitignore files in parent dirs up to and including repo-root
+// 			(ie only dirs that are an ancestor)
+// 		- read .git/info/exclude for repo
+//  - NOTE: the dir you're passing in directly could be a dir that is ignored
+//      by a parent's .gitignore or the per-repo exclude file; in this
+//      case, we still process the dir's .gitignore file and, possibly, but
+//      not it's sub-dirs
+//
+// 	- without recursive flag:
+// 		- read .gitignore in start dir
+// 		- read all .gitignore files in parent dirs up to and including repo-root
+// 			(ie only dirs that are an ancestor)
+// 		- read .git/info/exclude for repo
+//
+// - parsing a plain dir: .gitignore files are ignored
+//
+// In all cases any dirs matched by a previously read
+// .gitignore are skipped, unless it's the path (ie directly
+// supplied by the user).
+func ParseGitIgnores(path string, recursive bool) (returnPs []gitignore.Pattern, repoRootPath string, err error) {
+	// We need to parse .gitignore files from the root of the git repo to correctly identify ignored files
+	var fs billy.Filesystem
+	var ps, newPs []gitignore.Pattern
+
+	// Normalise to path (or directory containing path if it's a file)
+	path, err = getNormalisedDir(path)
+	if err != nil {
+		return nil, "", err
+	}
+
+	repo, err := git.PlainOpenWithOptions(path, &git.PlainOpenOptions{DetectDotGit: true})
+	if err != nil && !errors.Is(err, git.ErrRepositoryNotExists) {
+		return ps, "", err
+	}
+
+	// not in a git repo; do not read .gitignore files
+	// (and ignore recursive setting)
+	if errors.Is(err, git.ErrRepositoryNotExists) {
+		return ps, path, nil
+	}
+
+	// inside a git repo
+
+	repoRootPath, err = getRepoRootPath(repo)
+	if err != nil {
+		return ps, "", err
+	}
+
+	pathAbs, err := filepath.Abs(path)
+	if err != nil {
+		return ps, "", err
+	}
+
+	// path WRT git-root
+	pathRel, err := filepath.Rel(repoRootPath, pathAbs)
+	if err != nil {
+		return ps, "", err
+	}
+
+	fs = osfs.New(repoRootPath)
+
+	// Read parent dirs
+
+	// read the per-repo info/exclude file
+	newPs, err = readIgnoreFile(fs, []string{"."}, infoExcludeFile)
+	if err != nil {
+		return ps, "", err
+	}
+	ps = append(ps, newPs...)
+
+	// Read parent's dirs up to git-root
+	//
+	if pathAbs != repoRootPath {
+		newPs, err = readIgnoreFilesFromParents(fs, pathRel, repoRootPath)
+		ps = append(ps, newPs...)
+	}
+
+	// Read children's subdirs
+	//
+	switch {
```

Comment:
nit: I think it'll be clearer to just use a if else here.

Reply:
done